### PR TITLE
upgrade: reject formulae checked in earlier levels

### PR DESCRIPTION
**Pull request by amyspark**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [X] Have you successfully run `brew tests` with your changes locally?

-----

@MikeMcQuaid  -- I finally hit that depth level issue you had previously told me about in #6397, when upgrading the packages below:

```
amalia@Sakura:~ $ brew outdated
ffmpeg (4.2.1) < 4.2.1_1
gdk-pixbuf (2.38.2) < 2.40.0
gtk+3 (3.24.12) < 3.24.12_1
libsoup (2.68.1) < 2.68.2
vala (0.46.2) < 0.46.3
x264 (r2917) < r2917_1
yarn (1.19.0) < 1.19.1
```

This pull request ensures that only new kegs are scanned in each level. I've attached the stacktrace in a separate file, because it is TOO BIG.
[massive-log.txt](https://github.com/Homebrew/brew/files/3720870/massive-log.txt)

CAVEAT: I know `brew style` will fail -- I'm happy to get any suggestions in order to fix that double-block chain.

### CODE REVIEW

File: Library/Homebrew/cmd/upgrade.rb [hunk 1]
```diff
@@ -266,6 +266,8 @@ def broken_dependents(kegs, formulae)
         dependents = kegs.select do |keg|
           keg.runtime_dependencies
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/upgrade.rb]:
```suggestion
          keg = keg.runtime_dependencies
```

File: Library/Homebrew/cmd/upgrade.rb [hunk 2]
```diff
@@ -266,6 +266,8 @@ def broken_dependents(kegs, formulae)
...
              .any? { |d| d["full_name"] == formula.full_name }
+        end.reject do |keg|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/upgrade.rb]:
```suggestion
          keg unless scanned.include?(keg)
```

File: Library/Homebrew/cmd/upgrade.rb [hunk 3]
```diff
@@ -266,6 +266,8 @@ def broken_dependents(kegs, formulae)
...
+          scanned.include?(keg)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/upgrade.rb]:
```suggestion
```

File: Library/Homebrew/cmd/upgrade.rb [hunk 4]
```diff
@@ -290,7 +292,9 @@ def broken_dependents(kegs, formulae)
           descendants << f
         end
 
-        descendants_to_reinstall, descendants_pinned = broken_dependents(kegs, descendants)
+        scanned << dependents
+
+        descendants_to_reinstall, descendants_pinned = broken_dependents(kegs, descendants, scanned)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/upgrade.rb]: Maybe pass `scanned.uniq` (unless you make it a `Set`)?

File: Library/Homebrew/cmd/upgrade.rb [hunk 5]
```diff
@@ -290,7 +292,9 @@ def broken_dependents(kegs, formulae)
           descendants << f
         end
 
-        descendants_to_reinstall, descendants_pinned = broken_dependents(kegs, descendants)
+        scanned << dependents
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/upgrade.rb]: if `dependents` is an array this should be `scanned += dependents`

File: Library/Homebrew/cmd/upgrade.rb [hunk 6]
```diff
@@ -255,7 +255,7 @@ def upgradable_dependents(kegs, formulae)
     [formulae_to_upgrade, formulae_pinned]
   end
 
-  def broken_dependents(kegs, formulae)
+  def broken_dependents(kegs, formulae, scanned = [])
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/upgrade.rb]: What about `scanned = Set.new`?

**Comment by MikeMcQuaid** [comment]
Thanks again @amyspark!

**Comment by smoothdvd** [comment]
Still has this issue in 2.1.15.

**Comment by MikeMcQuaid** [comment]
@smoothdvd see https://github.com/Homebrew/brew/issues/6586

**Pull request merged**